SwanContents/swancontents/filemanager/eos/fileio.py

```diff
@@ -132,45 +132,46 @@ def read_notebook_from_local(path, encoding='utf-8', as_version=4):
-        if ('USERDATA_PATH' in os.environ and os.path.isdir(os.environ['USERDATA_PATH'])):
+        if 'USERDATA_PATH' in os.environ and os.path.isdir(os.environ['USERDATA_PATH']):
             # Define the filename for the local copy
             dirname, basename = os.path.split(os_path)
             local_fname = "-".join([os.environ['USER'], "nb", str(int(time.time())), dirname.replace(os.sep, "-"), basename])
             local_path = os.path.join(os.environ['USERDATA_PATH'], local_fname)
 
             # Write the notebook locally and check for consistency
             local_retry = 10
-            write_notebook_to_local(local_path, nb)
-            for i in range(0, local_retry):
-                read_nb = read_notebook_from_local(local_path)
-                if (nb == read_nb):
+            # Use asyncio.to_thread to offload blocking I/O operations
+            await asyncio.to_thread(write_notebook_to_local, local_path, nb)
+            for i in range(local_retry):
+                read_nb = await asyncio.to_thread(read_notebook_from_local, local_path)
+                if nb == read_nb:
                     break
                 else:
-                    #time.sleep(0.1*2**i)    # Backoff on retry (100ms to 51.2s)
-                    time.sleep(0.5)
-                    write_notebook_to_local(local_path, nb)
-
-        # In all cases, save on eos via fuse
-        with self.atomic_writing(os_path, encoding="utf-8") as f:
-            nbformat.write(
-                nb,
-                f,
-                version=nbformat.NO_CONVERT,
-                capture_validation_error=capture_validation_error,
-            )
-
-    @contextmanager
-    def atomic_writing(self, os_path, *args, **kwargs):
+                    await asyncio.to_thread(write_notebook_to_local, local_path, nb)
+
+        async with self.atomic_writing(os_path, encoding="utf-8") as f:
+            await asyncio.to_thread(nbformat.write, nb, f, version=nbformat.NO_CONVERT, capture_validation_error=capture_validation_error)
+
+    
+
+    @asynccontextmanager
+    async def atomic_writing(self, os_path, *args, **kwargs):
         """Overload the default atomic_writing to use a different write method
         From the original documentation:
         wrapper around atomic_writing that turns permission errors to 403.
         Depending on flag 'use_atomic_writing', the wrapper perform an actual atomic writing or
         simply writes the file (whatever an old exists or not)"""
 
         if self.use_atomic_writing:
-            with self.perm_to_403(os_path):
-                with atomic_writing(os_path, *args, log=self.log, **kwargs) as f:
-                    yield f
+            self.perm_to_403(os_path)
+
+            # Run the synchronous atomic_writing function in a new thread
+            file = await asyncio.to_thread(None, atomic_writing, os_path, *args, **kwargs)
+
+            try:
+                yield file
+            except:
+                raise
```

Comment:
Why this construct? Why not just yielding `f`?

Reply:
I added the old method again, so everything is as before (to be aligned with upstream).